File: public/tools/client/server/ServiceChannel.ts
import {EventEmitter} from "../../utils/EventEmitter";
import {Queue} from "../../utils/Queue";
import {Channel} from "../../gtp3/Channel";
import {Server} from "./Server";
import {Service} from "../../utils/Service";

/**
 * A queued outgoing message.
 * If a ServiceChannel is closed when calling request() or send(), open() will be automatically called.
 * The message is then saved in a queue until the channel is ready for sending it.
 */
type QueuedMessage<T> = {
	key: string;
	data: any;
	flags: number;
	silent: boolean;
	deferred: PromiseResolver<T>
}

/**
 * Interface of a function that can be decorated by ServiceChannel.Dispatch
 */
export type DispatchHandler = (payload: any) => void;

/**
 * Interface of the table used by ServiceChannel.Dispatch.
 * A single dispatch table is shared by every decorated function in an object.
 */
export type DispatchTable = Map<string, Map<string, DispatchHandler>>;

/**
 * A ServiceChannel
 * Like Channel, but automatically managed
 */
export class ServiceChannel extends EventEmitter {
	/**
	 * Outgoing queue of messages
	 */
	private queue = new Queue<QueuedMessage<any>>();

	/**
	 * Channel is open for the first time
	 */
	private first_open: boolean = true;

	/**
	 * The channel is currently being open
	 */
	private open_pending: boolean = false;

	/**
	 * The channel is closed
	 */
	private closed: boolean = true;

	/**
	 * The underlying Channel object
	 */
	private channel: Channel;

	/**
	 * Constructor
	 * @param server    The server object
	 * @param name      Name of this channel
	 * @param lazy      If lazy, the underlying channel will not be created until the first request or message
	 */
	constructor(private server: Server, private name: string, lazy: boolean) {
		super();
		if (!lazy) {
			this.open();
		}
	}

	/**
	 * Open the service channel.
	 * Called automatically when doing request or sending message on a closed channel.
	 */
	public async open() {
		if (this.channel || this.open_pending) return;

		this.closed = false;
		this.open_pending = true;

		try {
			// Open the actual channel
			const chan = await this.server.openChannel(this.name);
			this.channel = chan;

			// Emit a reset event if this is not the first time this channel is opened
			if (this.first_open) {
				this.first_open = false;
			} else {
				this.emit("reset");
			}

			this.emit("state", true);

			// Pipe every event to this emitter except reset and close
			chan.pipe(this, "!", "reset", "close");

			// Handle close
			chan.on("close", () => {
				this.channel = null;
				this.emit("state", false);
				if (this.closed) return;
				this.open();
			});

			// Flush queue
			while (!this.queue.empty()) {
				const item = this.queue.dequeue();
				if (item.deferred) {
					this.channel.request(item.key, item.data, item.flags, item.silent).then(
						(success) => item.deferred.resolve(success),
						(failure) => item.deferred.reject(failure)
					);
				} else {
					this.channel.send(item.key, item.data, item.flags);
				}
			}
		} catch (e) {
			throw new Error(`Failed to open ${this.name} channel: ${e}`);
		} finally {
			this.open_pending = false;
		}
	}

	/**
	 * Close the service channel.
	 * Unlink Channel.close, a service channel can be re-opened after being closed.
	 * @param code      Close code
	 * @param reason    Close reason
	 */
	public close(code?: number, reason?: string) {
		this.closed = true;
		if (this.channel)
			this.channel.close(code, reason);
	}

	/**
	 * Send a request on the service channel.
	 * Similar to Channel.request, but will open the service channel if lazy and not yet opened.
	 * @param key               Message type
	 * @param data              Message payload
	 * @param flags             Initial flags for the message
	 * @param silent            If silent, this request will not trigger the activity indicator
	 * @returns {Promise<T>}    The request result
	 */
	public request<T>(key: string, data?: any, flags?: number, silent?: boolean): Promise<T> {
		if (this.channel) {
			return this.channel.request<T>(key, data, flags, silent);
		} else {
			const deferred = Promise.defer<T>();
			this.queue.enqueue({ key, data, flags, silent, deferred });
			this.open();
			return deferred.promise;
		}
	}

	/**
	 * Send a message on the service channel.
	 * Similar to Channel.send, but will open the service channel if lazy and not yet opened.
	 * @param key   Message type
	 * @param data  Message payload
	 * @param flags Initial flags for the message
	 */
	public send(key: string, data?: any, flags?: number) {
		if (this.channel) {
			this.channel.send(key, data, flags);
		} else {
			this.queue.enqueue({ key, data, flags, silent: false, deferred: null });
			this.open();
		}
	}

	/**
	 * Construct a decorator function that will invoke the decorated function when
	 * a given message is received on the channel.
	 * @param source    The service channel property name
	 * @param message   The message type to catch
	 * @param splat     If true, the message content must be an array and this array will
	 *                  be used as arguments list to the function instead of the array itself.
	 */
	static Dispatch(source: string, message: string, splat: boolean = false) {
		return (target: Service, property: string) => {
			let table = Reflect.getMetadata<DispatchTable>("servicechannel:dispatch", target);
			if (!table) {
				table = new Map();
				Reflect.defineMetadata("servicechannel:dispatch", table, target);
			}

			let mapping = table.get(source);
			if (!mapping) {
				mapping = new Map();
				table.set(source, mapping);
			}

			const fn = <DispatchHandler> (<any> target)[property];
			mapping.set(message, splat ? function (p) { fn.apply(this, p); } : fn);
		};
	}

	/**
	 * Construct a decorator function that will update the decorated property with the
	 * current state of the service channel given as `source`.
	 * @param source    The service channel property name
	 */
	static ReflectState(source: string) {
		return (target: Service, property: string) => {
			let list = Reflect.getMetadata<[string, string][]>("servicechannel:state", target);
			if (!list) {
				list = [];
				Reflect.defineMetadata("servicechannel:state", list, target);
			}
			list.push([source, property]);
		};
	}
}

File: client/src/gt/services/base/ServiceChannel.scala
package gt.services.base

import boopickle.DefaultBasic._
import gt.{App, Server}
import gtp3.{Channel, PickledPayload}
import org.scalajs.dom.console
import rx.{Rx, Var}
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.{Future, Promise}
import scala.language.implicitConversions
import scala.util.{Failure, Success}

/**
  * A service channel that handle automatic mangement of a GTP3 channel
  * required by a service.
  *
  * @param tpe      the type of channel to request
  * @param lzy      if true, the channel will not be opened until required
  * @param delegate the delegate used to handle incoming messages
  */
class ServiceChannel(tpe: String, lzy: Boolean, delegate: Delegate) {
	/** The channel instance */
	private[this] var channel: Channel = null

	/** Current channel status */
	private[this] val openStatus: Var[Boolean] = false

	/** A read-only view of the current status of this service channel */
	val open: Rx[Boolean] = openStatus

	/** Queue of pending messages of queries during channel opening */
	private[this] val queue = scalajs.js.Array[Channel => Unit]()

	/** Is true if the channel is currently being opened */
	private[this] var openPending = false

	/**
	  * Performs eager opening if the channel is not lazy.
	  * Automatically called when the owner channel is activated.
	  */
	private[services] def eagerOpen(): Unit = if (!lzy) lazyOpen()

	/**
	  * Lazily open the channel when it is first required.
	  */
	private[services] def lazyOpen(): Unit = {
		if (channel != null || openPending) return
		openPending = true

		Server.openChannel(tpe).andThen {
			case _ => openPending = false
		}.onComplete {
			case Success(chan) =>
				channel = chan
				openStatus := true

				chan.onMessage ~> (delegate.receiveMessage _).tupled
				chan.onClose ~> { case (code, reason) =>
					channel = null
					openStatus := false
				}

				while (queue.length > 0) {
					queue.shift().apply(chan)
				}

			case Failure(e) =>
				console.error(s"Failed to open service channel $tpe: ${ App.formatException(e) }")
		}
	}

	/**
	  * Closes the service channel
	  */
	private[services] def close(): Unit = if (channel != null) {
		channel.close()
		channel = null
		openStatus := false
	}

	/**
	  * Executes an action once the service channel is available.
	  *
	  * @param action the action to execute, taking the channel as argument
	  * @tparam T the type of the result of the action
	  * @return the future returned by the action
	  */
	private def withChannel[T](action: Channel => Future[T]): Future[T] = {
		if (channel != null) action(channel)
		else {
			val promise = Promise[T]()
			queue.push(chan => promise.completeWith(action(chan)))
			lazyOpen()
			promise.future
		}
	}

	/**
	  * Sends a message without arguments to the server.
	  *
	  * Equivalent to calling `send(message, data)` with Unit as data.
	  *
	  * @param message the message to send
	  */
	def send(message: String): Unit = send(message, ())

	/**
	  * Sends a message to the server with some arguments.
	  *
	  * @param message the message to send
	  * @param data    argument to the message
	  * @tparam P the type of the arguments,
	  *           an implicit Pickler must exist for this type
	  */
	def send[P: Pickler](message: String, data: P): Unit = {
		withChannel { chan => Future.successful(chan.send(message, data)) }
	}

	/**
	  * Sends a request to the server.
	  *
	  * This overload does not accept arguments and will send Unit as a subsitute.
	  * It is not possible to perform a silent request with this overload. If required,
	  * the same behavior can be obtained by calling `request(req, data)` with `()` as
	  * data.
	  *
	  * @param req the request key
	  * @return a Pending request that will be executed when a callback is defined or a
	  *         concrete, unpickled, type is specified with the .as[T] method.
	  */
	def request(req: String): PendingRequest[Unit] = request(req, (), false)

	/**
	  * Sends a request to the server with some arguments.
	  *
	  * @param req    the request key
	  * @param data   arguments to the request
	  * @param silent if true, the request will not trigger the activity indicator
	  * @tparam P the type of data sent, an implicit Pickler must exists for this type
	  * @return a PendingRequest that will be executed when a callback is defined or a
	  *         concrete, unpickled, type is specified with the .as[T] method.
	  */
	def request[P](req: String, data: P, silent: Boolean = false): PendingRequest[P] = {
		PendingRequest(req, data, silent)
	}

	/**
	  * A pending request ready to be sent.
	  *
	  * This class is used to move the implicit pickler out of the request call.
	  * This allow chaining a call to request without interfering with the implicit
	  * argument waiting to be given.
	  *
	  * In case the expected return type is Future[PickledPayload], the implicit
	  * conversion defined in the companion object kicks-in and ensure that the
	  * request is properly sent.
	  *
	  * Most of the time however, client code will either call .as[T] or one of the
	  * .apply methods.
	  *
	  * @param req    capture of the request action
	  * @param data   capture of the request data
	  * @param silent capture of the silent flag of the request
	  * @tparam P the type of the payload sent to the server
	  */
	case class PendingRequest[P](req: String, data: P, silent: Boolean) {
		/** Executes the request and return a future of its result */
		@inline def future(implicit p: Pickler[P]): Future[PickledPayload] = withChannel { chan =>
			chan.request(req, data, silent)
		}

		/** Executes the request and return a future of the requested type (after automatic unpickling) */
		@inline def as[T: Pickler](implicit p: Pickler[P]): Future[T] = future.map(_.as[T])

		/** Executes the request and forward the result to the given handler function */
		@inline def apply[T1: Pickler, R](fn: T1 => R)(implicit p: Pickler[P]): Future[R] = {
			future.map { pp => fn(pp.as[T1]) } andThen {
				case Failure(e) => console.error("Uncaught exception during message dispatch:\n\n" + App.formatException(e))
			}
		}

		/** Executes the request and forward the result to the given handler function */
		@inline def apply[T1: Pickler, T2: Pickler, R](fn: (T1, T2) => R)(implicit p: Pickler[P]): Future[R] = {
			apply(fn.tupled)
		}

		/** Executes the request and forward the result to the given handler function */
		@inline def apply[T1: Pickler, T2: Pickler, T3: Pickler, R](fn: (T1, T2, T3) => R)(implicit p: Pickler[P]): Future[R] = {
			apply(fn.tupled)
		}

		/** Executes the request and forward the result to the given handler function */
		@inline def apply[T1: Pickler, T2: Pickler, T3: Pickler, T4: Pickler, R](fn: (T1, T2, T3, T4) => R)(implicit p: Pickler[P]): Future[R] = {
			apply(fn.tupled)
		}

		/** Executes the request and forward the result to the given handler function */
		@inline def apply[T1: Pickler, T2: Pickler, T3: Pickler, T4: Pickler, T5: Pickler, R](fn: (T1, T2, T3, T4, T5) => R)(implicit p: Pickler[P]): Future[R] = {
			apply(fn.tupled)
		}
	}

	object PendingRequest {
		/** Automatically executes a pending request in case a Future[PickledPayload] is expected */
		@inline implicit def AutoExec[P](r: PendingRequest[P])(implicit p: Pickler[P]): Future[PickledPayload] = r.future
	}
}
